### ISSUE TITLE
Homebrew package dependencies should support version ranges and not always depend on latest versions

### ISSUE THREAD

**Issue by vfonic** [Homebrew/homebrew-core#86425]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

<details>
<summary>brew config</summary>

```shell
HOMEBREW_VERSION: 3.2.14
ORIGIN: https://github.com/Homebrew/brew
HEAD: 4e6919b73444e9f9e02ad81d8afc4bbd97533567
Last commit: 5 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: b3db614229214629fc5bf9554d37a734da8c88a2
Core tap last commit: 84 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: /usr/local/bin/code -n -w
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: dodeca-core 64-bit kabylake
Clang: 13.0.0 build 1300
Git: 2.33.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.6-x86_64
CLT: 13.0.0.0.1.1630607135
Xcode: 13.0
```

</details>

<details>
<summary>brew doctor</summary>

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  proj@7
```

</details>

### 

- [X] I ran `brew update` and am still able to reproduce my issue.
- [ ] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

I have pinned python@3.9 package because I didn't want to run into any issues after `brew upgrade` and it was also time consuming to upgrade python...and, on top of it all, it was just working™.

I wanted to run `brew upgrade` to upgrade some other packages. However, I ran into issue where the package I'm trying to upgrade depends on python@3.9: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/gobject-introspection.rb#L25
I have python@3.9 installed. However, there's a newer version of python@3.9 available (which I don't want to install for the reasons above).

Why isn't it possible to install the package that depends on python@3.9 if I don't have the very latest python@3.9 version installed?

This can happen even in the situations where I have compatible python@3.9 version, but I still cannot upgrade (or insatll for that matter) the package.

Am I missing something? Did I misunderstand anything?

### What happened (include all command output)?

```
> brew upgrade
Error: Not upgrading 7 pinned packages:
postgresql 13.4, postgresql@12 12.8, postgis 3.1.4, node 16.10.0, python@3.8 3.8.12, node@14 14.18.0, python@3.9 3.9.7
==> Upgrading 14 outdated packages:
deno 1.11.2 -> 1.14.2
gdal 3.3.0_2 -> 3.3.2_2
libuv 1.41.0 -> 1.42.0
pango 1.48.5 -> 1.48.10
harfbuzz 2.8.1 -> 3.0.0
glib 2.68.3 -> 2.70.0
netpbm 10.86.22 -> 10.86.24
qt 6.1.1 -> 6.1.3
gobject-introspection 1.68.0_1 -> 1.70.0
yarn 1.22.10 -> 1.22.15
numpy 1.21.0 -> 1.21.2
graphviz 2.47.3 -> 2.49.1
poppler 21.06.1 -> 21.10.0
imagemagick 7.1.0-2 -> 7.1.0-9
Error: You must `brew unpin python@3.9` as installing glib requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing imagemagick requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing numpy requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing gdal requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing libuv requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing gobject-introspection requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing harfbuzz requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing pango requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing netpbm requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing qt requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9 node` as installing yarn requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing graphviz requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing poppler requires the latest version of pinned dependencies
Error: You must `brew unpin python@3.9` as installing deno requires the latest version of pinned dependencies
```

### What did you expect to happen?

Something along the lines of:

```
Error: You must `brew unpin python@3.9` as installing glib requires the latest version of pinned dependencies
Or you can run `brew upgrade glib --force` to force upgrading
```

Or:

```
Or you can run `brew upgrade glib --skip-pinned` to force upgrading
```

Or:

```
Or you can run `brew upgrade glib --skip-pinned-check` to force upgrading
```

Or it could show that as a warning or as a prompt:

```
glib depends on python@3.9 which is pinned. Run `brew unpin python@3.9` or continue installing with python@3.9 pinned? [Y/n]
```

### Step-by-step reproduction instructions (by running `brew` commands)

This is difficult. :)

1. Install python@3.9 package (I don't know how to install older version of a package)
2. Wait for new python@3.9 version to be released (or install older version if you know how)
3. `brew pin python@3.9`
4. (Try to) install a package that depends on python@3.9.

Expected:

Package installs with warning or prompts user to force-install

Actual:

Package installation fails with error "You must unpin python@3.9"

**Comment by SMillerDev** [comment]
> Why isn't it possible to install the package that depends on python@3.9 if I don't have the very latest python@3.9 version installed?

Because they aren't tested with that version. For python this might work, but not everyone does versioning cleanly.

---

This is a feature request, not a bug, which we only allow in the form of pull requests. Closing

**Comment by carlocab** [comment]
This is, unfortunately, really impractical to implement. We don't have a way of verifying that a given non-latest-version of `foo` actually works with `bar`. Building infrastructure to do this will very quickly cause us to exceed the limited CI capacity that we have.

**Comment by vfonic** [comment]
Thank you for such a prompt reply! Now I understand why this repo has so few open issues. :) 👍 

> This is a feature request, not a bug, which we only allow in the form of pull requests. Closing

Ah, I missed the memo on how to submit feature request. It is a bit odd to have to open a PR, especially for this kind of feature request (which would also use CI credits) and I don't even know where to start. I assume a good start would be: check if this feature would be accepted before starting to work on it?
Adding a hint here about how to submit a feature request would help:
https://github.com/Homebrew/homebrew-core/issues/new/choose

> We don't have a way of verifying that a given non-latest-version of `foo` actually works with `bar`.

That's why I proposed `--force` or similar solution. This would allow for a much simpler solution that would go a long way: We don't recommend upgrading, but you can try.

There could even be this scenario:

`bar@1` 1.0.0 released on Oct 4th
`foo` (with `bar@1` as dependency) 1.0.0 released on Oct 5th
`foo` 2.0.0 released on Oct 15th
`bar@1` 1.0.1 released on Oct 30th

User timeline:
`brew install foo` on Oct 10th
`brew pin bar@1` on Oct 10th
`brew upgrade` on Nov 1st

Would this user get an error on Nov 1st saying that `bar@1` is pinned, even though update for `bar@1` was released after `foo` update and we know that `foo` is compatible with an older (currently installed) version of `bar@1`?

I think, from a user's (user experience) perspective, this feature, supported in any way, would really improve the experience. Basically, without being able to do something like `brew upgrade foo --force`, it really feels like the "Apple way": "We know better than you". And even when this is not true as shown in my example above.

In my example with `python@3.9`, I don't see the added value of upgrading python version when absolutely everything is working with the currently installed version of `python@3.9`. Why would I risk the chance of breaking my current setup when the only (visibly) added benefit is: I'd have latest version, as-in a higher version number than before. 🤷‍♂️

**Comment by vfonic** [comment]
I checked PRs and I couldn't find a single feature request there. 🤷‍♂️ 

Should I open a PR for this issue or is this considered a WONTFIX with no additional discussion needed?

Thanks!

**Comment by carlocab** [comment]
I'm not opposed to `--force` continuing an install despite pinned dependencies, and will review a PR that implements this (with the warnings adjusted accordingly after having passed `--force` as well). Other maintainers may have a different opinion than I do, however.

**Comment by SMillerDev** [comment]
> Why would I risk the chance of breaking my current setup when the only (visibly) added benefit is: I'd have latest version, as-in a higher version number than before. 

Because you don't always see security updates. It takes effort to release something, you can trust that developers have a reason for it.

> I checked PRs and I couldn't find a single feature request there. 🤷‍♂️

That's because those changes will need to be made in Homebrew/brew.

> Would this user get an error on Nov 1st saying that bar@1 is pinned, even though update for bar@1 was released after foo update and we know that foo is compatible with an older (currently installed) version of bar@1?

Yes, because bar 1.0.1 might be needed to resolve an issue in foo 2.0.0. Which the pin is blocking.

**Comment by vfonic** [comment]
> Because you don't always see security updates.

I'm aware of that. I choose not to update, and I don't see the need to get the latest security update for a tool that's only (if ever) used on my machine.

> It takes effort to release something, you can trust that developers have a reason for it.

There could be many reasons for releasing a software update and I don't think that everyone should always upgrade to the latest version immediately. This also takes time and effort.

> Yes, because bar 1.0.1 might be needed to resolve an issue in foo 2.0.0. Which the pin is blocking.

There could even be this scenario:

`bar@1` 1.0.0 released on Oct 4th
`foo` (with `bar@1` as dependency) 1.0.0 released on Oct 5th
`foo` 2.0.0 released on Oct 15th
`bar@1` 1.0.1 released on Oct 30th

User timeline:
`brew install foo` on Oct 10th
`brew pin bar@1` on Oct 10th
`brew upgrade` on Nov 1st

But the user would get this error, even if there are users who ran `brew upgrade` before Oct 30th (installed `foo` 2.0.0 with `bar@1` 1.0.0 and had everything working? It seems odd that some users would be able to upgrade the package if they ran the upgrade before `bar@1` 1.0.1 was released, while other users are not able to get the same upgrade of `foo` package because `bar@1` 1.0.1 was released. 🤔 
We're just assuming here that `foo` _maybe_ wouldn't work with `bar@1` 1.0.0.

**Comment by SMillerDev** [comment]
> There could be many reasons for releasing a software update and I don't think that everyone should always upgrade to the latest version immediately.

Then brew might not be the package manager for you, since this is the philosophy that brew's update model is based on.

>  We're just assuming here that foo maybe wouldn't work with bar@1 1.0.0.

We're assuring that foo 2.0.0 works with bar 1.0.1 by running it through CI. Any other configuration is unsupported because we won't know if it works. Because in this ideal scenario it could work, but what if bar updates from version 20211004 to 20211030? Then there is no way to see if they are compatible. There simply isn't enough certainty in version numbers for brew to make these kind of assumptions, so we don't.

**Comment by vfonic** [comment]
> We're assuring that foo 2.0.0 works with bar 1.0.1 by running it through CI.

Didn't the CI run when `foo` 2.0.0 was released as well?

The CI runs every time when there's a new package version, am I correct?

So it will run in all of these situations:
`bar@1` 1.0.0 released on Oct 4th ==> CI runs with `bar@1` 1.0.0, and no `foo`
`foo` (with `bar@1` as dependency) 1.0.0 released on Oct 5th ==> CI runs with `bar@1` 1.0.0, and `foo` 1.0.0
`foo` 2.0.0 released on Oct 15th ==> CI runs with `bar@1` 1.0.0, and `foo` 2.0.0 <===== here
`bar@1` 1.0.1 released on Oct 30th ==> CI runs with `bar@1` 1.0.1, and `foo` 2.0.0

We just had the CI check the compatibility of these two versions where it says "here" above.
Or am I missing something?

I'll stop here. We have different opinions and I feel we are starting to go in circles. I appreciate the time you took to look into this issue! Thanks!
